lib/kubeclient/common.rb

```diff
@@ -484,5 +485,26 @@ def http_options(uri)
 
       options.merge(@socket_options)
     end
+
+    def use_default_gcp_token
+      scopes =  ['https://www.googleapis.com/auth/cloud-platform']
+      authorization = Google::Auth.get_application_default(scopes)
+
+      authorization.apply({})
+
+      authorization.access_token
```

Comment:
This is not a lot of code, so I'm wondering if kubeclient needs to include it at all.
- Does this cover most uses with GCP, or will people need other variations?
- https://github.com/Shopify/kubernetes-deploy/pull/88 says it's valid for 1 hour.  Is that tunable?  **Should a long-lived Client object be able to renew it?**

The alternative I'm thinking about is:
- Document example code in README.
- Those directly passing `auth_options:` just write this or similar code and pass the token.
- For those reading a config file, Kubeclient::Config could have a hook to plug your own handling of 'auth-provider'.  E.g. `auth-provider: gcp` would call `fetch_auth_provider_options_gcp` method one could override in a subclass.

See also my other comment about one Config / many `context` calls / many Client instances / one or many tokens.  Perhaps for gcp you better take matters into your own hands and not call `config.context.auth_options` at all?  (for that, Config should let user access the raw `auth-provider` portion of config)

Reply:
I'm sure there are other cases with GCP but I can think of the two:

1. User has authorized `gcloud` for their GCP project (and by extension `kubectl`). In that case, the token is only good for an hour as I pointed out in #210. I tried to address this in #211 but I couldn't get it to work for more than an hour and I'm not sure if we can.

2. This is on a GCP server (GKE, GCE, App Engine, Cloud Functions), where Google has installed "default credentials". That's what this PR addresses. (You can always [install default credentials on your local dev machine](https://cloud.google.com/sdk/gcloud/reference/auth/application-default/login) as "long-running" credentials.)

I see your point that this could just be documented. It would be really nice if `kubeclient` just worked out-of-the-box for `gcloud`-configured machines (e.g. development machine) and GCE/GKE machines. I can see that sometimes users are going to want to modify the scope and may eventually want to use specific keys for role-based-access (rather than authorizing the default credentials on a machine for this permission). So maybe there's a place for basic "out-of-the-box" support and anything beyond that is documented.

It does make more sense to use `Config` than `Client`. However, the shopify solution depends on the presence of `auth-provider.name` in the configuration file but there is no such file in a GCP server. So it wouldn't solve for that case.